subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ExecuteStep.java

```diff
@@ -28,10 +28,14 @@
     @Override
     public Result execute(C context) {
         UnitOfWork work = context.getWork();
-        ExecutionOutcome outcome = context.getInputChanges()
-            .map(inputChanges -> determineOutcome(work.execute(inputChanges, context), inputChanges.isIncremental()))
-            .orElseGet(() -> determineOutcome(work.execute(null, context), false));
-        return () -> Try.successful(outcome);
+        try {
+            ExecutionOutcome outcome = context.getInputChanges()
+                .map(inputChanges -> determineOutcome(work.execute(inputChanges, context), inputChanges.isIncremental()))
+                .orElseGet(() -> determineOutcome(work.execute(null, context), false));
+            return () -> Try.successful(outcome);
```

Comment:
This is a lot like using `Try.ofFailable`, only that we catch a `Throwable` and not `Exception`. Do you think it makes sense to add a method catching `Throwable` to `Try`, something like `Try.ofFailableUserCode`?

Reply:
I think `Try` is much too low-level to handle a concept like user code. We could have some utility somewhere to capture the execution of user code in a uniform way, though.

It would be good to clarify how this works, too. For example, I think we should let `Error`s propagate.

Alas, all of this is outside of the scope of this PR.